src/lib/comms/sol-lwm2m-security.c

```diff
@@ -0,0 +1,470 @@
+        return -ENOENT;
+    }
+
+    SOL_VECTOR_FOREACH_IDX (&obj_ctx->instances, instance, i) {
+        r = read_resources(lwm2m_client, obj_ctx, instance, res, 4,
+            SECURITY_SECURITY_MODE,
+            SECURITY_SERVER_ID,
+            SECURITY_PUBLIC_KEY_OR_IDENTITY,
+            SECURITY_IS_BOOTSTRAP);
+        SOL_INT_CHECK_GOTO(r, < 0, err_clear);
+
+        //If -ENOENT when reading SECURITY_PUBLIC_KEY_OR_IDENTITY or
+        // SECURITY_SECURITY_MODE != SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY
+        if (res[2].data_len == 0 ||
+            res[0].data[0].content.integer != SOL_LWM2M_SECURITY_MODE_PRE_SHARED_KEY) {
+            clear_resource_array(res, sol_util_array_size(res));
+            continue;
+        }
+
+        if (res[1].data[0].content.integer == server_id ||
+            (res[3].data[0].content.b && (UINT16_MAX == server_id))) {
+            size_t psk_id_len;
+
+            if (res[2].data[0].content.blob->size > SOL_DTLS_PSK_ID_LEN) {
+                SOL_WRN("The PSK ID '%.*s' is %zu-bytes long;"
+                    " expecting a PSK ID at most %d-bytes long",
+                    SOL_STR_SLICE_PRINT(sol_str_slice_from_blob(res[2].data[0].content.blob)),
+                    res[2].data[0].content.blob->size, SOL_DTLS_PSK_ID_LEN);
+                r = -EINVAL;
+                goto err_clear;
+            }
+
+            psk_id_len = res[2].data[0].content.blob->size;
+
+            memcpy(id, res[2].data[0].content.blob->mem, psk_id_len);
+            clear_resource_array(res, sol_util_array_size(res));
+
+            return (ssize_t)psk_id_len;
+        }
+
+        clear_resource_array(res, sol_util_array_size(res));
+    }
+
+    return -ENOENT;
+
+err_clear:
+    clear_resource_array(res, sol_util_array_size(res));
+    return r;
+}
+
+static void
+sol_lwm2m_security_del_full(struct sol_lwm2m_security *security,
+    enum lwm2m_entity_type type)
+{
+    SOL_NULL_CHECK(security);
+
+    switch (type) {
+    case LWM2M_CLIENT:
+        ((struct sol_lwm2m_client *)(security->entity))->security = NULL;
+        break;
+    case LWM2M_SERVER:
+        ((struct sol_lwm2m_server *)(security->entity))->security = NULL;
+        break;
+    case LWM2M_BOOTSTRAP_SERVER:
+        ((struct sol_lwm2m_bootstrap_server *)(security->entity))->security = NULL;
+        break;
+    }
+
+    sol_util_clear_memory_secure(security, sizeof(*security));
+    free(security);
+}
+
+static struct sol_lwm2m_security *
+sol_lwm2m_security_add_full(void *entity, enum lwm2m_entity_type type,
+    enum sol_lwm2m_security_mode sec_mode)
+{
+    struct sol_lwm2m_security *security;
+    struct sol_socket *socket_dtls = NULL;
+    int r = 0;
+
+    ssize_t (*get_psk_cb)(const void *creds, struct sol_str_slice id,
+        char *psk, size_t psk_len) = NULL;
+    ssize_t (*get_id_cb)(const void *creds, struct sol_network_link_addr *addr,
+        char *id, size_t id_len) = NULL;
```

Comment:
[2], and not for the first time, but during the whole development of LWM2M code :(